### PR TITLE
Add docs for Data Prepper dissect processor

### LINKED ISSUE #5106: [DOC] Documentation for Data Prepper dissect processor
**What do you want to do?**
 
- [ ] Request a change to existing documentation
- [x] Add new documentation
- [ ] Report a technical problem with the documentation
- [ ] Other 

**Tell us about your request.** Provide a summary of the request and all versions that are affected.
Requesting documentation for a new dissect processor in Data Prepper. This is documented in a README file in data prepper repo: https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-plugins/dissect-processor/README.md

The processor will be part of the Data Prepper 2.5 release, planned for early October.

**What other resources are available?** Provide links to related issues, POCs, steps for testing, etc.

### PR DESCRIPTION
**Pull request by oeyh**
### Description
Add docs for dissect processor

### Issues Resolved
Resolves #5106

### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: _data-prepper/pipelines/configuration/processors/dissect.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+nav_order: 52
+---
+
+# dissect
+
+The `dissect` processor extracts values from an event and map them to individual fields based on user defined Dissect patterns. The processor is well-suited for field extraction from log messages with a known structure. 
+
+## Basic usage
+
+For an example using dissect processor, create the following `pipeline.yaml`.
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/dissect.md:15]:
```suggestion
For an example of using the `dissect` processor, create the following `pipeline.yaml`:

```

File: _data-prepper/pipelines/configuration/processors/dissect.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      format: "json"
+  processor:
+    - dissect:
+        map:
+          log: "%{Date} %{Time} %{Log_Type}: %{Message}"
+  sink:
+    - stdout:
+```
+
+Then create the following file named `logs_json.log` and replace the `path` in the file source of your `pipeline.yaml` with the path of a file containing this JSON data:
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/dissect.md:31]:
```suggestion
Then create the following file, named `logs_json.log`, and replace the `path` in the file source of your `pipeline.yaml` with the path of a file containing the following JSON data:
```

File: _data-prepper/pipelines/configuration/processors/dissect.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+```
+{"log": "07-25-2023 10:00:00 ERROR: error message"}
+```
+
+The `dissect` processor will retrieve the fields (`Date`, `Time`, `Log_Type`, and `Message`) from the `log` message, , based on the pattern `%{Date} %{Time} %{Type}: %{Message}` configured in the pipeline.
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/dissect.md:37]:
```suggestion
The `dissect` processor will retrieve the fields (`Date`, `Time`, `Log_Type`, and `Message`) from the `log` message, based on the pattern `%{Date} %{Time} %{Type}: %{Message}` configured in the pipeline.
```

File: _data-prepper/pipelines/configuration/processors/dissect.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+When you run Data Prepper with this `pipeline.yaml` passed in, you should see the following standard output.
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/dissect.md:39]:
```suggestion
When you run Data Prepper the following`pipeline.yaml` passed in, you should receive  the following standard output:
```

File: _data-prepper/pipelines/configuration/processors/dissect.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+}
+```
+
+## Configuration
+
+You can configure the `dissect` processor with the following options.
+
+| Option | Required | Type | Description |
+| :--- | :--- | :--- | :--- |
+| `map` | Yes | Map | Defines the dissect patterns for specific keys. For details on how to define fields in the dissect pattern, see [Field notations](#field-notations) |
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/dissect.md:57]:
```suggestion
| `map` | Yes | Map | Defines the `dissect` patterns for specific keys. For details on how to define fields in the dissect pattern, see [Field notations](#field-notations) |
```

File: _data-prepper/pipelines/configuration/processors/dissect.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+| `target_types` | No | Map | Specifies the data types for extract fields. Valid options are `integer`, `double`, `string`, and `boolean`. By default, all fields are of `string` type. |
+| `dissect_when` | No | String | Specifies a condition for performing the dissect operation using a [Data Prepper expression]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/expression-syntax/). If specified, dissect only when the expression evaluates to true. |
+
+### Field notations
+
+You can define a dissect patterns with the following types of fields.
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/dissect.md:63]:
```suggestion
You can define `dissect` patterns with the following types of fields.
```

File: _data-prepper/pipelines/configuration/processors/dissect.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+#### Normal field
+
+A field without a suffix or prefix. The field will be directly added to the output Event. The format is `%{field_name}`.
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/dissect.md:67]:
```suggestion
A field without a suffix or prefix. The field will be directly added to the output event. The format is `%{field_name}`.
```

File: _data-prepper/pipelines/configuration/processors/dissect.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+#### Skip field
+
+A field that will not be put in the event. The format is `%{}` or `%{?field_name}`.
+
+#### Append field
+
+A field that will be combined with other fields. To append multiple values and put the final value in the field, we can use + before the field name in the dissect pattern. The format is `%{+field_name}`.
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/dissect.md:75]:
```suggestion
A field that will be combined with other fields. To append multiple values and put the final value in the field, use `+` before the field name in the `dissect` pattern. The format is `%{+field_name}`. 
```

File: _data-prepper/pipelines/configuration/processors/dissect.md [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+#### Skip field
+
+A field that will not be put in the event. The format is `%{}` or `%{?field_name}`.
+
+#### Append field
+
+A field that will be combined with other fields. To append multiple values and put the final value in the field, we can use + before the field name in the dissect pattern. The format is `%{+field_name}`. 
+
+For example, with a pattern `%{+field_name}, %{+field_name}`, log message `"foo, bar"` will parse into `{"field_name": "foobar"}`.
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/dissect.md:77]:
```suggestion
For example, with the pattern `%{+field_name}, %{+field_name}`, log message `"foo, bar"` will parse into `{"field_name": "foobar"}`.
```

File: _data-prepper/pipelines/configuration/processors/dissect.md [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+We can also define the order the concatenation with the help of suffix `/<integer>`.
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/dissect.md:79]:
```suggestion
You can also define the order the concatenation with the help of suffix `/<integer>`. 
```

**Review by Naarcha-AWS — APPROVED** [review]

File: _data-prepper/pipelines/configuration/processors/dissect.md [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+You can define `dissect` patterns with the following types of fields.
+
+#### Normal field
+
+A field without a suffix or prefix. The field will be directly added to the output event. The format is `%{field_name}`.
+
+#### Skip field
+
+A field that will not be put in the event. The format is `%{}` or `%{?field_name}`.
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/dissect.md:72]:
```suggestion
A field that will not be included in the event. The format is `%{}` or `%{?field_name}`.
```

File: _data-prepper/pipelines/configuration/processors/dissect.md [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+#### Append field
+
+A field that will be combined with other fields. To append multiple values and put the final value in the field, use `+` before the field name in the `dissect` pattern. The format is `%{+field_name}`. 
+
+For example, with the pattern `%{+field_name}, %{+field_name}`, log message `"foo, bar"` will parse into `{"field_name": "foobar"}`.
+
+You can also define the order the concatenation with the help of suffix `/<integer>`.
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/dissect.md:80]:
```suggestion
You can also define the order of the concatenation with the help of the suffix `/<integer>`. 
```

File: _data-prepper/pipelines/configuration/processors/dissect.md [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+
+You can also define the order the concatenation with the help of suffix `/<integer>`. 
+
+For example, with a pattern `"%{+field_name/2}, %{+field_name/1}"`, log message `"foo, bar"` will parse into `{"field_name": "barfoo"}`.
+
+If the order is not mentioned, the append operation will take place in the order of fields specified in the dissect pattern. 
+
+#### Indirect field
+
+A field that uses the value from another field as field name. While defining a pattern, prefix the field with a `&` to assign the value found with this field to the value of another field found as the key.
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/dissect.md:88]:
```suggestion
A field that uses the value from another field as its field name. While defining a pattern, prefix the field with a `&` to assign the value found with this field to the value of another field found as the key.
```

Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/dissect.md:88]: I'm not following the second sentence. Can it be simplified?

File: _data-prepper/pipelines/configuration/processors/dissect.md [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+For example, with a pattern `"%{?field_name}, %{&field_name}"`, the log message `"foo, bar"` will parse into `{“foo”: “bar”}`. In the message, `foo` is captured from the skip field `%{?field_name}` serves as the key to the value captured from the field `%{&field_name}`.
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/dissect.md:90]: Same here. The second sentence is too complex. Let's simplify/clarify.

File: _data-prepper/pipelines/configuration/processors/dissect.md [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+#### Padded field
+
+A field with paddings to the right removed. `->` operator can be used as a suffix to a field to indicate that white spaces after this field can be ignored.
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/dissect.md:94]:
```suggestion
A field with the paddings to the right removed. The `->` operator can be used as a suffix to indicate that white spaces after this field can be ignored.
```

Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/dissect.md:94]: Does "spaces" need to be plural here?

**Review by natebower — CHANGES_REQUESTED** [review]
@Naarcha-AWS Requesting changes because I'll want to read lines 15, 40, 60, 88, and 90 again before approving. Please tag me when ready. Thanks!

File: _data-prepper/pipelines/configuration/processors/dissect.md [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+
+Then create the following file, named `logs_json.log`, and replace the `path` in the file source of your `pipeline.yaml` with the path of a file containing the following JSON data:
+
+```
+{"log": "07-25-2023 10:00:00 ERROR: error message"}
+```
+
+The `dissect` processor will retrieve the fields (`Date`, `Time`, `Log_Type`, and `Message`) from the `log` message, based on the pattern `%{Date} %{Time} %{Type}: %{Message}` configured in the pipeline.
+
+When you run Data Prepper the following`pipeline.yaml` passed in, you should receive  the following standard output:
```
[the remark this answers is not in the document]
Comment (Naarcha-AWS) [reply _data-prepper/pipelines/configuration/processors/dissect.md:40]: The code is just the output. Let's switch to, "After running the pipeline."

File: _data-prepper/pipelines/configuration/processors/dissect.md [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+
+## Configuration
+
+You can configure the `dissect` processor with the following options.
+
+| Option | Required | Type | Description |
+| :--- | :--- | :--- | :--- |
+| `map` | Yes | Map | Defines the `dissect` patterns for specific keys. For details on how to define fields in the dissect pattern, see [Field notations](#field-notations) |
+| `target_types` | No | Map | Specifies the data types for extract fields. Valid options are `integer`, `double`, `string`, and `boolean`. By default, all fields are of `string` type. |
+| `dissect_when` | No | String | Specifies a condition for performing the dissect operation using a [Data Prepper expression]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/expression-syntax/). If specified, dissect only when the expression evaluates to true. |
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/dissect.md:60]:
```suggestion
| `dissect_when` | No | String | Specifies a condition for performing the `dissect` operation using a [Data Prepper expression]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/expression-syntax/). If specified, the `dissect` operation will only run when the expression evaluates to true. |
```

File: _data-prepper/pipelines/configuration/processors/dissect.md (hunk 14 above)
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/dissect.md:90]:
```suggestion
For example, with a pattern `"%{?field_name}, %{&field_name}"`, the log message `"foo, bar"` will parse into `{“foo”: “bar”}`. In the log message, `foo` is captured from the skip field `%{?field_name}`. `foo` then serves as the key to the value captured from the field `%{&field_name}`.
```

File: _data-prepper/pipelines/configuration/processors/dissect.md [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+nav_order: 52
+---
+
+# dissect
+
+The `dissect` processor extracts values from an event and maps them to individual fields based on user-defined `dissect` patterns. The processor is well suited for field extraction from log messages with a known structure. 
+
+## Basic usage
+
+To create a `dissect` processor, create the following `pipeline.yaml` file:
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/dissect.md:15]:
```suggestion
To use the `dissect` processor, create the following `pipeline.yaml` file:
```

File: _data-prepper/pipelines/configuration/processors/dissect.md [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+
+## Configuration
+
+You can configure the `dissect` processor with the following options.
+
+| Option | Required | Type | Description |
+| :--- | :--- | :--- | :--- |
+| `map` | Yes | Map | Defines the `dissect` patterns for specific keys. For details on how to define fields in the `dissect` pattern, see [Field notations](#field-notations). |
+| `target_types` | No | Map | Specifies the data types for extract fields. Valid options are `integer`, `double`, `string`, and `boolean`. By default, all fields are of the `string` type. |
+| `dissect_when` | No | String | Specifies a condition for performing the `dissect` operation using a [Data Prepper expression]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/expression-syntax/). If specified, the `dissect` operation will only run when the expression evaluates to true. |
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/dissect.md:60]: Should "true" be in code font here?

File: _data-prepper/pipelines/configuration/processors/dissect.md [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+
+You can also define the order of the concatenation with the help of the suffix `/<integer>`. 
+
+For example, with a pattern `"%{+field_name/2}, %{+field_name/1}"`, log message `"foo, bar"` will parse into `{"field_name": "barfoo"}`.
+
+If the order is not mentioned, the append operation will occur in the order of the fields specified in the `dissect` pattern. 
+
+#### Indirect field
+
+A field that uses the value from another field as its field name. While defining a pattern, prefix the field with a `&` to assign the value found with this field to the value of another field found as the key.
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/dissect.md:88]:
```suggestion
A field that uses the value from another field as its field name. When defining a pattern, prefix the field with a `&` to assign the value found in the field as the key in the key-value pair.
```

**Review by natebower — APPROVED** [review]
@Naarcha-AWS Approved with the implementation of these comments/changes. Thanks!

**Pull request merged**